[1주차] 간단한 계산기

timothypark33 opened the pull request:
# 와일드 백엔드 1주차 과제
## 목적
Java Http 통신으로 계산기 API 구현하기

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **신규 기능**
  - HTTP 서버 구현으로 GET 및 POST 요청 처리 가능.
  - "/calculations" 경로에 대한 POST 요청으로 기본 산술 연산 수행.
  - "/calculations" 경로에 대한 GET 요청으로 계산 목록 조회 가능.
  - JSON 데이터 직렬화 및 역직렬화 기능 향상.
  - 계산 요청 및 응답을 위한 DTO 클래스 추가.
  - 계산 저장 및 조회를 위한 리포지토리 인터페이스 및 구현 추가.

- **변경 사항**
  - 서버 시작 시 상태 메시지 출력.
  - 테스트를 위한 새로운 계산기 테스트 클래스 추가.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

timothypark33 committed:
main 내에 사용 코드를 run()에 옮겨서 분리

timothypark33 committed:
http 서버 기본 형태 구현

timothypark33 committed:
단순 계산기 API 기능 구현

timothypark33 committed:
계산기 기능을 RequestHandler에 분리

timothypark33 committed:
RequestHandler에서 아래 함수 추가
getRequestKey
getRequestContent
sendResponseContent

timothypark33 committed:
requestKey 단위로 코드 분리

ahastudio commented on app/src/main/java/com/example/demo/presentation/CalculationCreateResource.java:
```diff
@@ -0,0 +1,27 @@
+        int number1 = Integer.parseInt(splitedRequestBody[0]);
+        int number2 = Integer.parseInt(splitedRequestBody[2]);
+        String operation = splitedRequestBody[1];
+        String responseContent = "";
+
+        if (operation.equals("+")) {
+            responseContent = String.valueOf(number1 + number2);
+        } else if (operation.equals("-")) {
+            responseContent = String.valueOf(number1 - number2);
+        } else if (operation.equals("*")) {
+            responseContent = String.valueOf(number1 * number2);
+        } else if (operation.equals("/")) {
+            responseContent = String.valueOf(number1 / number2);
+        } else {
+            responseContent = "Error";
+        }
+        return responseContent;
```
임시 변수를 쓰지 말고 바로 리턴하세요. 대부분의 경우 `else`는 불필요합니다.

```suggestion
        if (operation.equals("+")) {
            return String.valueOf(number1 + number2);
        }
        if (operation.equals("-")) {
            return String.valueOf(number1 - number2);
        }
        if (operation.equals("*")) {
            return String.valueOf(number1 * number2);
        }
        if (operation.equals("/")) {
            return String.valueOf(number1 / number2);
        }

        // 많은 경우, 여기서 예외를 던집니다.
        return "Error";
```

timothypark33 replied:
피드백 감사합니다

timothypark33 committed:
chore: 불필요한 else 제거

timothypark33 committed:
사칙연산을 application로 분리

timothypark33 committed:
JSON으로 입력받고 응답

ahastudio commented on app/build.gradle.kts:
```diff
@@ -9,6 +9,8 @@ repositories {
 }
 
 dependencies {
+    implementation("com.fasterxml.jackson.core:jackson-databind:2.18.0")
+
```
@coderabbitai // 이건 뭐라고 생각해?

https://mvnrepository.com/artifact/com.fasterxml.jackson.core/jackson-databind/2.18.0

ahastudio commented on app/src/main/java/com/example/demo/application/OperatorKind.java:
```diff
@@ -0,0 +1,25 @@
+package com.example.demo.application;
+
+public enum OperatorKind {
```
다형성을 이용할 경우엔 이를 통해 enum을 제거합니다. enum은 static과 함께 객체지향에서 지양해야 할 것 중 하나입니다.
지금 코드에서 enum이 가져다 주는 이점이 뭐가 있는지, 오히려 단점이 되는 건 뭐가 있는지 고민해 보세요.

timothypark33 replied:
연산자가 정해져있기 때문에 변하지 않은 타입을 주기위해 enum을 사용했지만 실제 코드 작성에서도 불편을 느꼈습니다.
수식에서 연산자를 가져오는데 있어서 String에서 OperationKind 클래스로 변환하는데 있어서 불편했습니다 변환하는 코드도 작성해야하니 더 복잡해졌습니다.

ENUM을 사용하지 않고 연산자를 구분하는 것이 더 구현/가독성에 모두 편한것같습니다.

ahastudio commented on app/src/main/java/com/example/demo/presentation/CalculationCreateResource.java:
```diff
@@ -0,0 +1,35 @@
+                            result
+                    )
+            );
+        } catch (ArrayIndexOutOfBoundsException | IllegalArgumentException e) {
```
예외를 굳이 catch하지 않아도 됩니다. 지금은 예외를 잡아서 그냥 문자열을 주기 때문에 HTTP Status Code 200을 주면서 JSON이 아닌 다른 걸 주는 더 큰 문제를 발생시키게 됩니다. 서버가 에러를 내는 것과, 서버에서 잘 준 것 같았지만 사실은 아니라서 클라이언트가 죽는 것은 완전히 다릅니다. 사용자에게 문제를 일으키지 않아야 합니다.

timothypark33 committed:
GET /calculations 로 계산 히스토리 반환

timothypark33 commented on app/src/main/java/com/example/demo/presentation/RequestHandler.java:
```diff
@@ -0,0 +1,63 @@
+        }
+        response += "\n";
+        sendResponseContent(exchange, response);
+    }
```
사칙연산 코드를 application 폴더에 구분해서 구현했습니다.
RequestHandler 클래스의 handle 메서드에서 API Request Path와 HTTP Method을 구분해서 Resource 클래스를 반환합니다.

timothypark33 replied:
"연산 유형에 대해 열거형(enum) 또는 상수를 사용하여 코드의 유지보수성을 개선하세요."
너가 해준 피드백은 오히려 코드를 더 복잡하게 만드는것같아
단순히 String으로 구분해도 충분하다고 판단돼

ENUM을 사용하면 반드시 아래처럼 변환과정이 필요해서 불편하다고 판단돼
Operation operation = Operation.fromString(parts[1]);

그럼에도 불구하고 ENUM을 사용을 권유하고 싶다면 추가적인 설명을 듣고싶어

timothypark33 commented on app/src/main/java/com/example/demo/application/OperatorMultiply.java:
```diff
@@ -0,0 +1,8 @@
+package com.example.demo.application;
+
+public class OperatorMultiply implements Operator {
+    @Override
+    public int calculate(int number1, int number2) {
+        return 0;
+    }
+}
```
1번과 3번의 예시를 보여줘

2번은 문서화를 어느정도 진행해야할까? 코드가 단순해서 문서화가 필요가 없을 수 있을것같지만
나중에 코드가 복잡해진다면 미리 조금이라도 양식에 갖춰서 작성해두는게 좋을것같아

timothypark33 replied:
단위 테스트 는 어디에 모아두고 실행시킬 수 있지?

timothypark33 commented on app/src/main/java/com/example/demo/presentation/HomeGetResource.java:
```diff
@@ -0,0 +1,10 @@
+package com.example.demo.presentation;
+
+public class HomeGetResource extends ResourceMethodHandler {
+    public final static String KEY = "GET /";
```
public enum HttpMethod {
    GET,
    POST,
    PUT,
    DELETE
    // 필요한 다른 HTTP 메서드들...
}

이 코드를  폴더에 구분해서 작성하는게 좋아보이는데 폴더이름은 뭘로 하는게 좋을까?

timothypark33 committed:
"계산 프로그램"에 맞지않는 Resource 제거

timothypark33 committed:
ENUM인 OperatorKind.java 제거로 복잡성 해결

timothypark33 committed:
OperationMultiply 계산식 정상화

timothypark33 committed:
OperatorMinus 줄바꿈

timothypark33 commented on app/src/main/java/com/example/demo/application/OperatorMultiply.java:
```diff
@@ -0,0 +1,8 @@
+    @Override
+    public int calculate(int number1, int number2) {
+        return 0;
+    }
```
해결했음

timothypark33 committed:
OperatorDivide 잠재적 문제 예외처리

timothypark33 committed:
1. CalculationRequestDto 연산자 예외처리
2. JAVADOC 주석추가

timothypark33 commented on app/src/main/java/com/example/demo/presentation/dto/CalculationRequestDto.java:
```diff
@@ -0,0 +1,19 @@
+package com.example.demo.presentation.dto;
+
+public record CalculationRequestDto(
+    int number1,
+    int number2,
+    String operator
+){
```
개선 완료

timothypark33 committed:
레코드는 자동으로 각 필드에 대한 접근자 메서드를 생성합니다.
따라서 getNumber1(), getNumber2(), getOperator() 메서드를 제거.

timothypark33 commented on app/src/main/java/com/example/demo/presentation/CalculationCreateResource.java:
```diff
@@ -0,0 +1,35 @@
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+public class CalculationCreateResource extends ResourceMethodHandler {
+    public final static String KEY = "POST /calculations";
+    public final Calculator calculator = new Calculator();
+    private final ObjectMapper objectMapper = new ObjectMapper();
```
public final static String KEY 로 접근하나

private final static String KEY
public static String getKey() {return KEY;} 로 접근하나

똑같이 접근하는것데 왜 캡슐화까지 해야해?
캡슐화 하더라도 외부에서 얼마든지 getKey로 접근할 수 있다고 생각해

ahastudio replied:
ResourceMethodHandler에 `getKey` 추상 메서드를 추가해서 쓰면 유용합니다. 밖에서 put할 때 어떤 클래스인지 몰라도 되고, DI를 한다면 더욱 강력하죠.

timothypark33 committed:
유효성 검사 로직 개선

timothypark33 committed:
getCalculations 메서드에서 불변 리스트 반환 고려

timothypark33 committed:
record getter 사용

timothypark33 committed:
sendResponseContent 메서드 개선 반영

timothypark33 committed:
getRequestContent 메서드 개선 반영

timothypark33 committed:
getRequestKey 메서드 개선 반영

timothypark33 committed:
RequestHandler handle 메서드 개선 반영

timothypark33 committed:
openapi.yaml 추가

timothypark33 committed:
feat: spring framework 사용을 위한 라이브러리 설치
org.springframework:spring-core:6.1.13
org.springframework:spring-context:6.1.13

timothypark33 committed:
spring-setup: App.java에
ApplicationContext 생성

timothypark33 committed:
httpServer 사용 로직을 Server.java에 분리해서 사용

timothypark33 committed:
requestHandler를 멤버 변수로 사용

timothypark33 committed:
@Bean으로 Bean 관리

timothypark33 committed:
@ComponentScan과 @Component로 Bean 관리
Server와 RequestHandler에 적용

timothypark33 committed:
ResourceMethodHandler에 getKey를 추가

timothypark33 committed:
ResourceMethodHandler 상속 받은 클래스들을 @Component로 Bean관리
calculationCreateResource
calculationListResource

timothypark33 committed:
CalculationCreateResource, CalculationListResource -> ~Hander로 이름 변경

timothypark33 committed:
CalculationCreateHandler에서 사용하기 위해
ObjectMapper를 @Bean으로 관리
Calculator를 @Component로 관리

timothypark33 committed:
RequestHandler에서 사용하는 클래스 이름 변경
CalculationCreateResource -> CalculationCreateHandler
CalculationListResource -> CalculationListHandler

timothypark33 committed:
@Component로 Bean관리에 따른 CalculationListHandler 생성자 코드 변경

timothypark33 committed:
CalculationRepositoryf를 @Component로 Bean관리

timothypark33 committed:
테스트 코드 작성을 위한 라이브러리 추가
"org.assertj:assertj-core:3.26.3"
"org.junit.jupiter:junit-jupiter:5.10.0"

timothypark33 committed:
사칙연산 테스트 코드 작성

timothypark33 committed:
파일이름 변경 CalcaulationRepository -> InMemoryCalculationRepository
CalcaulationRepository는 interface로 사용

timothypark33 committed:
FakeCalculationRepository는
CalculationRepository Mocking한 것. 테스트 코드에 사용

timothypark33 committed:
불필요한 import 제거
